README
list.c: 線形リスト演習用
paint.c: 配列とtail index で履歴保存を実装したもの
paint_linearlist.c: 線形リストで履歴保存を実装する雛形
yamanote.txt: 最近話題の山手線


history.txt
line 2 2 100 2
line 10 15 30 35
circle 40 20 8
line 10 10 20 20
line 10 20 140 20
circle 100 20 10
circle 100 20 5
circle 100 20 3
line 10 30 100 30
pen
line 10 10 30 10
circle 40 20 5
pen @
circle 70 15 6
pen O
line 140 10 140 30


paint.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

// array for history recording
const int max_history = 5;
const int bufsize = 1000;

char history[max_history][bufsize];

// Structure for canvas
typedef struct
{
  int width;
  int height;
  char **canvas;
  char pen;
} Canvas;

Canvas *init_canvas(int width,int height, char pen)
{
  Canvas *new = (Canvas *)malloc(sizeof(Canvas));
  new->width = width;
  new->height = height;
  new->canvas = (char **)malloc(width * sizeof(char *));

  char *tmp = (char *)malloc(width*height*sizeof(char));
  memset(tmp, ' ', width*height*sizeof(char));
  for (int i = 0 ; i < width ; i++){
    new->canvas[i] = tmp + i * height;
  }

  new->pen = pen;
  return new;
}

void reset_canvas(Canvas *c)
{
  const int width = c->width;
  const int height = c->height;
  memset(c->canvas[0], ' ', width*height*sizeof(char));
}

void print_canvas(FILE *fp, Canvas *c)
{
  const int height = c->height;
  const int width = c->width;
  char **canvas = c->canvas;
  
  // 上の壁
  fprintf(fp,"+");
  for (int x = 0 ; x < width ; x++)
    fprintf(fp, "-");
  fprintf(fp, "+\n");

  // 外壁と内側
  for (int y = 0 ; y < height ; y++) {
    fprintf(fp,"|");
    for (int x = 0 ; x < width; x++){
      const char c = canvas[x][y];
      fputc(c, fp);
    }
    fprintf(fp,"|\n");
  }
  
  // 下の壁
  fprintf(fp, "+");
  for (int x = 0 ; x < width ; x++)
    fprintf(fp, "-");
  fprintf(fp, "+\n");
  fflush(fp);
}

void free_canvas(Canvas *c)
{
  free(c->canvas[0]); //  for 2-D array free
  free(c->canvas);
  free(c);
}

void rewind_screen(FILE *fp,unsigned int line)
{
  fprintf(fp,"\033[%dA",line);
}

void clear_command(FILE *fp)
{
  fprintf(fp,"\033[2K");
}
void clear_screen(FILE *fp)
{
  fprintf(fp, "\033[2J");
}

int max(const int a, const int b)
{
  return (a > b) ? a : b;
}

void draw_line(Canvas *c, const int x0, const int y0, const int x1, const int y1)
{
  const int width = c->width;
  const int height = c->height;
  char pen = c->pen;
  
  const int n = max(abs(x1 - x0), abs(y1 - y0));
  c->canvas[x0][y0] = pen;
  for (int i = 1; i <= n; i++) {
    const int x = x0 + i * (x1 - x0) / n;
    const int y = y0 + i * (y1 - y0) / n;
    if ( (x >= 0) && (x<width) && (y >= 0) && (y < height))
      c->canvas[x][y] = pen;
  }
}

void save_history(const char *filename, const int history_size)
{
  const char *default_history_file = "history.txt";
  if (filename == NULL)
    filename = default_history_file;
  
  FILE *fp;
  if ((fp = fopen(filename, "w")) == NULL) {
    fprintf(stderr, "error: cannot open %s.\n", filename);
    return;
  }
  
  for (int i = 0; i < history_size; i++) {
    fprintf(fp, "%s", history[i]);
  }

  printf("saved as \"%s\"\n", filename);

  fclose(fp);
}

// Interpret and execute a command
//   return value:
//     0, normal commands such as "line"
//     1, unknown or special commands (not recorded in history[])
//     2, quit
int interpret_command(const char *command, int *hsize, Canvas *c)
{
  char buf[bufsize];
  strcpy(buf, command);
  buf[strlen(buf) - 1] = 0; // remove the newline character at the end

  const char *s = strtok(buf, " ");

  // The first token corresponds to command
  if (strcmp(s, "line") == 0) {
    int x0, y0, x1, y1;
    x0 = 0; y0 = 0; x1 = 0; y1 = 0; // initialize
    x0 = atoi(strtok(NULL, " "));
    y0 = atoi(strtok(NULL, " "));
    x1 = atoi(strtok(NULL, " "));
    y1 = atoi(strtok(NULL, " "));
    draw_line(c,x0, y0, x1, y1);
    return 0;
  }
  
  if (strcmp(s, "save") == 0) {
    s = strtok(NULL, " ");
    save_history(s, *hsize);
    return 1;
  }

  if (strcmp(s, "undo") == 0) {
    reset_canvas(c);
    for (int i = 0; i < *hsize - 1; i++) {
      interpret_command(history[i], NULL,c);
    }
    (*hsize)--;
    return 1;
  }

  if (strcmp(s, "quit") == 0) {
    return 2;
  }

  printf("error: unknown command.\n");

  return 1;
}

int main()
{
  const int width = 70;
  const int height = 40;
  char pen = '*';
  
  FILE *fp;
  char buf[bufsize];
  fp = stdout;
  Canvas *c = init_canvas(width,height, pen);
  
  int hsize = 0;  // history size
  while (hsize < max_history) {

    printf("%d > ", hsize);
    fgets(buf, bufsize, stdin);

    const int r = interpret_command(buf, &hsize,c);
    if (r == 2) break;
    if (r == 0) {
      strcpy(history[hsize], buf);
      hsize++;
    }

    print_canvas(fp, c);
    rewind_screen(fp, height+3); // rewind the screen to command input
    clear_command(fp); // clear the previous command
  }

  clear_screen(fp);
  free_canvas(c);
  fclose(fp);

  return 0;
}


paint1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <assert.h>

const int bufsize = 1000;

struct history
{
  char *command;
  struct history *next;
};
typedef struct history History;

// Structure for canvas
typedef struct {
  int width;
  int height;
  char **canvas;
  char pen;
} Canvas;

History *undo(History *h, Canvas *c);
int interpret_command(const char *, History *, Canvas *);

History *init_history() {
  History *begin = NULL;
  return begin;
}

int history_size(History *h) {
  int hsize = 0;
  History *p;
  for (p = h ; p != NULL ; p = p->next)
    hsize++;
  return hsize;
}

History *push_history(History *begin, char *command) {
  // Create a new element
  History *new = (History *)malloc(sizeof(History));
  char *c = (char *)malloc(strlen(command) + 1);
  strcpy(c, command);
  new->command = c;
  new->next = NULL;

  if (begin == NULL) return new;

  History *current = begin;
  while(current->next != NULL){
    current = current->next;
  }
  current->next = new;
  return begin;
}

History *undo(History *h, Canvas *c) {
  // 基本的にはpop_back を利用する。ただしループが進むついでにinterpret_command する

  // 現状履歴がないなら何もする必要がない
  // "No history" とでも表示する...
  if (h == NULL) {
    printf("No history\n");
    return NULL;
  }

  History *p = h;
  History *q = h->next;
  // 現在描画コマンド一つであれば、それを実行する必要はないのでそのまま消して終わり
  if (q == NULL){
    free(p->command);
    free(p);
    return NULL;
  }

  // q->next が NULLまでポインタを順次たどる
  while(q->next != NULL){
    // 更新作業のまえにinterpret_commandする (キャンバス更新)
    // 線形リストに記録されているコマンドでは履歴は操作されないので、NULLとしておく
    // interpret_command 内にも undo があるが問題ない（これを再帰関数という）
    interpret_command(p->command, NULL, c);

    p = q;
    q = p->next;
  }
  // このループを抜けた時点でqは終端だが、pに終端の一つ前が格納されている
  // このpはまだ実行されていないのでまず実行する
  // その後、削除すべきqを処理
  interpret_command(p->command, NULL, c);
  free(q->command);
  free(q); 
  p->next = NULL; //pが次の終端となるように指定

  return h;//
  
}


Canvas *init_canvas(int width,int height, char pen) {
  Canvas *new = (Canvas *)malloc(sizeof(Canvas));
  new->width = width;
  new->height = height;
  new->canvas = (char **)malloc(width * sizeof(char *));

  char *tmp = (char *)malloc(width*height*sizeof(char));
  memset(tmp, ' ', width*height*sizeof(char));
  for (int i = 0 ; i < width ; i++){
    new->canvas[i] = tmp + i * height;
  }

  new->pen = pen;
  return new;
}

void reset_canvas(Canvas *c) {
  const int width = c->width;
  const int height = c->height;
  memset(c->canvas[0], ' ', width*height*sizeof(char));
}

void print_canvas(FILE *fp, Canvas *c) {
  const int height = c->height;
  const int width = c->width;
  char **canvas = c->canvas;
  
  // 上の壁
  fprintf(fp,"+");
  for (int x = 0 ; x < width ; x++)
    fprintf(fp, "-");
  fprintf(fp, "+\n");

  // 外壁と内側
  for (int y = 0 ; y < height ; y++) {
    fprintf(fp,"|");
    for (int x = 0 ; x < width; x++){
      const char c = canvas[x][y];
      fputc(c, fp);
    }
    fprintf(fp,"|\n");
  }
  
  // 下の壁
  fprintf(fp, "+");
  for (int x = 0 ; x < width ; x++)
    fprintf(fp, "-");
  fprintf(fp, "+\n");
  fflush(fp);
}

void free_canvas(Canvas *c) {
  free(c->canvas[0]); //  for 2-D array free
  free(c->canvas);
  free(c);
}

void rewind_screen(FILE *fp,unsigned int line) {
  fprintf(fp,"\033[%dA",line);
}

void clear_command(FILE *fp) {
  fprintf(fp,"\033[2K");
}
void clear_screen(FILE *fp) {
  fprintf(fp, "\033[2J");
}

int max(const int a, const int b) {
  return (a > b) ? a : b;
}

void draw_line(Canvas *c, const int x0, const int y0, const int x1, const int y1) {
  const int width = c->width;
  const int height = c->height;
  char pen = c->pen;
  
  const int n = max(abs(x1 - x0), abs(y1 - y0));
  c->canvas[x0][y0] = pen;
  for (int i = 1; i <= n; i++) {
    const int x = x0 + i * (x1 - x0) / n;
    const int y = y0 + i * (y1 - y0) / n;
    if ( (x >= 0) && (x<width) && (y >= 0) && (y < height))
      c->canvas[x][y] = pen;
  }
}

void draw_rectangle(Canvas *c, const int x0, const int y0, const int x1, const int y1) {
  for (int i = y0; i < y1; i++) {
    draw_line(c,x0,i,x1,i);
  }
}

void draw_circle(Canvas *c, const int x0, const int y0, const int r) {
  const int width = c->width;
  const int height = c->height;
  char pen = c->pen;
  if ( (x0-r >= 0) && (x0+r <= width) && (y0-r >= 0) && (y0+r <= height)) {
    for (int y = y0-2*r; y < y0+2*r; y++) {
      for (int x = x0-3*r; x < x0+3*r; x++) {
        double d = sqrt(pow(x-x0,2)/4 + pow(y-y0,2));
        if ((int) d == r) {
          c->canvas[x][y] = pen;
        }
      }
    }
  }
}

void save_history(const char *filename, History *h) {
  const char *default_history_file = "history.txt";
  if (filename == NULL)
    filename = default_history_file;
  
  FILE *fp;
  if ((fp = fopen(filename, "w")) == NULL) {
    fprintf(stderr, "error: cannot open %s.\n", filename);
    return;
  }
  
  for (History *p = h ; p != NULL ; p = p->next){
    fprintf(fp,"%s",p->command);
  }

  printf("saved as \"%s\"\n", filename);

  fclose(fp);
}

// Interpret and execute a command
//   return value:
//     0, normal commands such as "line"
//     1, unknown or special commands (not recorded in history[])
//     2, quit
int interpret_command(const char *command, History *h, Canvas *c) {
  char buf[bufsize];
  strcpy(buf, command);
  buf[strlen(buf) - 1] = 0; // remove the newline character at the end

  const char *s = strtok(buf, " ");

  // The first token corresponds to command
  if (strcmp(s, "line") == 0) {
    int x0, y0, x1, y1;
    x0 = 0; y0 = 0; x1 = 0; y1 = 0; // initialize
    x0 = atoi(strtok(NULL, " "));
    y0 = atoi(strtok(NULL, " "));
    x1 = atoi(strtok(NULL, " "));
    y1 = atoi(strtok(NULL, " "));
    draw_line(c,x0, y0, x1, y1);
    return 0;
  }

  if (strcmp(s, "rect") == 0) {
    int x0, y0, x1, y1;
    x0 = 0; y0 = 0; x1 = 0; y1 = 0; // initialize
    x0 = atoi(strtok(NULL, " "));
    y0 = atoi(strtok(NULL, " "));
    x1 = atoi(strtok(NULL, " "));
    y1 = atoi(strtok(NULL, " "));
    draw_rectangle(c,x0,y0,x1,y1);
    return 0;
  }

  if (strcmp(s, "circle") == 0) {
    int x0, y0, r;
    x0 = 0; y0 = 0; r = 0;
    x0 = atoi(strtok(NULL, " "));
    y0 = atoi(strtok(NULL, " "));
    r = atoi(strtok(NULL, " "));
    draw_circle(c,x0,y0,r);
    return 0;
  }
  
  if (strcmp(s, "save") == 0) {
    s = strtok(NULL, " ");
    save_history(s, h);
    return 1;
  }

  if (strcmp(s, "undo") == 0) {
    reset_canvas(c);
    // do implementation of iterative interpret_command
    //
    h = undo(h,c);
    if (h == NULL){
      return 3;
    }
    return 1;
  }

  if (strcmp(s, "quit") == 0) {
    return 2;
  }

  printf("error: unknown command.\n");

  return 1;
}

int main() {
  const int width = 150;
  const int height = 40;
  char pen = '*';
  
  FILE *fp;
  char buf[bufsize];
  fp = stdout;
  Canvas *c = init_canvas(width,height, pen);
  
  int hsize = 0;  // history size

  History *h = init_history();

  print_canvas(fp, c);

  while (1) {
    hsize = history_size(h);

    printf("%d > ", hsize);
    fgets(buf, bufsize, stdin);

    const int r = interpret_command(buf, h,c);
    if (r == 2) break;
    if (r == 0) {
      h = push_history(h, buf);
    }

    print_canvas(fp, c);
    rewind_screen(fp, height+3); // rewind the screen to command input
    clear_command(fp); // clear the previous command
  }

  clear_screen(fp);
  free_canvas(c);
  fclose(fp);

  return 0;
}


sample/list.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

const int maxlen = 1000;

// A node which belongs to a linear list
// 線形リストの要素となる構造体
struct node 
{
  char *str;
  struct node *next;
};

// Add another name for struct by typedef
// typedef で Node型として使えるように宣言
typedef struct node Node;


// 線形リストの先頭に新しくデータを追加 (push) する関数
// 引数:
// - 現在の線形リストの先頭へのポインタ (begin)
// - 追加するべき文字列 (str)
//
// 出力:
// - 新しく追加されたNodeへのポインタ（新しく先頭になる）
//
Node *push_front(Node *begin, const char *str)
{
  // Create a new element
  // 追加するデータのために構造体とその中身を確保
  Node *p = (Node *)malloc(sizeof(Node));
  char *s = (char *)malloc(strlen(str) + 1);
  strcpy(s, str); // 文字列をコピーする
  p->str = s; //確保した構造体メンバーの文字列（=charポインタ）に代入
  p->next = begin; //引数で受け取った構造体を次の行き先に設定

  return p;  // Now the new element is the first element in the list
}

// 線形リストの先頭からデータを取り除く (pop) する関数
// 引数:
// - 現在の線形リストの先頭へのポインタ (begin)
//
// 出力:
// - 削除後に先頭になるNodeへのポインタ（もともと2番目だったもの）
//

Node *pop_front(Node *begin)
{
  // NULL のnext は当然指定できないので、現在、空の場合はとまる
  assert(begin != NULL); // Don't call pop_front() when the list is empty
  Node *p = begin->next; // 2番目のアドレスを確保（これをreturnする）

  free(begin->str);// 現在先頭の構造体の中身はfree
  free(begin);// 構造体自体もfreeする

  return p;
}

// 線形リストの末尾に新しくデータを追加 (push) する関数
// 引数:
// - 現在の線形リストの先頭へのポインタ (begin)
// - 追加するべき文字列 (str)
//
// 出力:
// - 線形リストの先頭へのポインタ（begin と同じ）
//

Node *push_back(Node *begin, const char *str)
{
  // 現状、空の場合は後ろには追加できないので、前に追加する
  if (begin == NULL) {   // If the list is empty
    return push_front(begin, str);
  }

  // Find the last element
  Node *p = begin;
  // インデックス以外でforやwhile ループを回す方法の一つ
  // 線形リストの場合はp->next で指定された次の構造体へのアドレスを順に辿ればいい
  // 次の要素がNULL になっているものを見つけるまで続ける
  while (p->next != NULL) {
    p = p->next;
  }
  // ループを抜けた時点でpは現在、末尾の構造体を指している
  
  // Create a new element
  // 追加する構造体用のメモリを確保する (このあたりはpush_frontの時と同じ)
  Node *q = (Node *)malloc(sizeof(Node));
  char *s = (char *)malloc(strlen(str) + 1);
  strcpy(s, str);
  q->str = s;
  q->next = NULL; //今回確保したものは末尾にあるので、nextがNULLをさすようにする

  // The new element should be linked from the previous last element
  // pからqにつなぐ
  p->next = q;

  return begin;
}

// 実習1: pop_back の実装
// 線形リストの先頭からデータを取り除く (pop) する関数
// 引数:
// - 現在の線形リストの先頭へのポインタ (begin)
//
// 出力:
// - 先頭になるNodeへのポインタ（begin と同じ）
//
Node *pop_back(Node *begin)
{
  // 終端を削除する場合は、一つ前を後から終端に指定 (p->next =NULL) する必要がある。
  
  // 先頭がNULLの場合は何もする必要がない
  if (begin == NULL){
    return begin; 
  }
  
  // 以下はそれ以外の場合
  // 二箇所 (現在p と次q) を覚えておく必要がある
  // 次q の次 (q->next) がNULL かどうかが終了判定 
  Node *p = begin;
  Node *q = p->next;

  // この時点でq==NULL (要素が一つだけだった場合) は p を削除して終わり
  if (q == NULL){
    free(p->str);
    free(p);
    return q;
  }
  // q->next がNULLまでポインタを順次たどる
  while (q->next != NULL){
    p = q;
    q = p->next;
  }
  // このループを抜けた時点でqは終端だが、pに終端の一つ前が格納されている
  // 削除すべきqを処理
  free(q->str);
  free(q); 
  p->next = NULL; //pが次の終端となるように指定
  
  return begin;
}


Node *remove_all(Node *begin)
{
  while ((begin = pop_front(begin))) 
    ; // Repeat pop_front() until the list becomes empty
  return begin;  // Now, begin is NULL
}

int main()
{
  Node *begin = NULL; // pointer to the first element in the list

  // Read all lines from stdin and store them in the list
  char buf[maxlen];
  while (fgets(buf, maxlen, stdin)) {
    //begin = push_front(begin, buf);
    begin = push_back(begin, buf); // Try this instead of push_front()
  }

  //begin = pop_front(begin);  // What will happen if you do this?
  //begin = pop_back(begin);   // What will happen if you do this?
 
  //begin = remove_all(begin); // What will happen if you do this?

  // Print all the strings stored in the list
  const Node *p;
  // ポインタを辿ってfor文を回している。先頭アドレスからスタートして、アドレスをp->next がさす次の構造体へ更新し、NULL (終端) に当たるまで続ける
  for (p = begin; p != NULL; p = p->next) {
    printf("%s", p->str);
  }
  
  return 0;
}


paint3.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <assert.h>

const int bufsize = 1000;

struct history
{
  char *command;
  struct history *next;
};
typedef struct history History;

// Structure for canvas
typedef struct {
  int width;
  int height;
  char **canvas;
  char pen;
} Canvas;

History *undo(History *h, Canvas *c);
int interpret_command(const char *, History *, Canvas *);

History *init_history() {
  History *begin = NULL;
  return begin;
}

int history_size(History *h) {
  int hsize = 0;
  History *p;
  for (p = h ; p != NULL ; p = p->next)
    hsize++;
  return hsize;
}

void show_history(History *h) {
  History *p = h;
  int i = 0;
  if (p == NULL) {
    printf("No history\n");
    return;
  }
  while (p != NULL) {
    printf("%d: %s", i, p->command);
    i++;
    p = p->next;
  }
}

History *push_history(History *begin, char *command) {
  // Create a new element
  History *new = (History *)malloc(sizeof(History));
  char *c = (char *)malloc(strlen(command) + 1);
  strcpy(c, command);
  new->command = c;
  new->next = NULL;

  if (begin == NULL) return new;

  History *current = begin;
  while(current->next != NULL){
    current = current->next;
  }
  current->next = new;
  return begin;
}

History *undo(History *h, Canvas *c) {
  // 基本的にはpop_back を利用する。ただしループが進むついでにinterpret_command する

  // 現状履歴がないなら何もする必要がない
  // "No history" とでも表示する...
  if (h == NULL) {
    printf("No history\n");
    return NULL;
  }

  History *p = h;
  History *q = h->next;
  // 現在描画コマンド一つであれば、それを実行する必要はないのでそのまま消して終わり
  if (q == NULL){
    free(p->command);
    free(p);
    return NULL;
  }

  // q->next が NULLまでポインタを順次たどる
  while(q->next != NULL){
    // 更新作業のまえにinterpret_commandする (キャンバス更新)
    // 線形リストに記録されているコマンドでは履歴は操作されないので、NULLとしておく
    // interpret_command 内にも undo があるが問題ない（これを再帰関数という）
    interpret_command(p->command, NULL, c);

    p = q;
    q = p->next;
  }
  // このループを抜けた時点でqは終端だが、pに終端の一つ前が格納されている
  // このpはまだ実行されていないのでまず実行する
  // その後、削除すべきqを処理
  interpret_command(p->command, NULL, c);
  free(q->command);
  free(q); 
  p->next = NULL; //pが次の終端となるように指定

  return h;//
  
}


Canvas *init_canvas(int width,int height, char pen) {
  Canvas *new = (Canvas *)malloc(sizeof(Canvas));
  new->width = width;
  new->height = height;
  new->canvas = (char **)malloc(width * sizeof(char *));

  char *tmp = (char *)malloc(width*height*sizeof(char));
  memset(tmp, ' ', width*height*sizeof(char));
  for (int i = 0 ; i < width ; i++){
    new->canvas[i] = tmp + i * height;
  }

  new->pen = pen;
  return new;
}

void reset_canvas(Canvas *c) {
  const int width = c->width;
  const int height = c->height;
  memset(c->canvas[0], ' ', width*height*sizeof(char));
}

void print_canvas(FILE *fp, Canvas *c) {
  const int height = c->height;
  const int width = c->width;
  char **canvas = c->canvas;
  
  // 上の壁
  fprintf(fp,"+");
  for (int x = 0 ; x < width ; x++)
    fprintf(fp, "-");
  fprintf(fp, "+\n");

  // 外壁と内側
  for (int y = 0 ; y < height ; y++) {
    fprintf(fp,"|");
    for (int x = 0 ; x < width; x++){
      const char c = canvas[x][y];
      fputc(c, fp);
    }
    fprintf(fp,"|\n");
  }
  
  // 下の壁
  fprintf(fp, "+");
  for (int x = 0 ; x < width ; x++)
    fprintf(fp, "-");
  fprintf(fp, "+\n");
  fflush(fp);
}

void free_canvas(Canvas *c) {
  free(c->canvas[0]); //  for 2-D array free
  free(c->canvas);
  free(c);
}

void rewind_screen(FILE *fp,unsigned int line) {
  fprintf(fp,"\033[%dA",line);
}

void clear_command(FILE *fp) {
  fprintf(fp,"\033[2K");
}
void clear_screen(FILE *fp) {
  fprintf(fp, "\033[2J");
}

int max(const int a, const int b) {
  return (a > b) ? a : b;
}

void draw_line(Canvas *c, const int x0, const int y0, const int x1, const int y1) {
  const int width = c->width;
  const int height = c->height;
  char pen = c->pen;
  
  const int n = max(abs(x1 - x0), abs(y1 - y0));
  c->canvas[x0][y0] = pen;
  for (int i = 1; i <= n; i++) {
    const int x = x0 + i * (x1 - x0) / n;
    const int y = y0 + i * (y1 - y0) / n;
    if ( (x >= 0) && (x<width) && (y >= 0) && (y < height))
      c->canvas[x][y] = pen;
  }
}

void draw_rectangle(Canvas *c, const int x0, const int y0, const int x1, const int y1) {
  for (int i = y0; i < y1; i++) {
    draw_line(c,x0,i,x1,i);
  }
}

void draw_circle(Canvas *c, const int x0, const int y0, const int r) {
  const int width = c->width;
  const int height = c->height;
  char pen = c->pen;
  if ( (x0-r >= 0) && (x0+r <= width) && (y0-r >= 0) && (y0+r <= height)) {
    for (int y = y0-2*r; y < y0+2*r; y++) {
      for (int x = x0-3*r; x < x0+3*r; x++) {
        double d = sqrt(pow(x-x0,2)/4 + pow(y-y0,2));
        if ((int) d == r) {
          c->canvas[x][y] = pen;
        }
      }
    }
  }
}

void save_history(const char *filename, History *h) {
  const char *default_history_file = "history.txt";
  if (filename == NULL)
    filename = default_history_file;
  
  FILE *fp;
  if ((fp = fopen(filename, "w")) == NULL) {
    fprintf(stderr, "error: cannot open %s.\n", filename);
    return;
  }
  
  for (History *p = h ; p != NULL ; p = p->next){
    fprintf(fp,"%s",p->command);
  }

  printf("saved as \"%s\"\n", filename);

  fclose(fp);
}

// ファイルからコマンド履歴をロードし、描写
// 最新のHistoryリストの先頭を返す
History *load(const char *command, History *h, Canvas *c) {
  char buf[bufsize];
  strcpy(buf, command);
  buf[strlen(buf) - 1] = 0;
  char *s = strtok(buf, " ");

  if (strcmp(s, "load") == 0) {
    char *filename = strtok(NULL, " ");
    char *default_history_file = "history.txt";
    if (filename == NULL)
      filename = default_history_file;
    FILE *fp;
    if ((fp = fopen(filename, "r")) == NULL) {
      fprintf(stderr, "error: cannot open %s.\n", filename);
      return NULL;
    }

    char *command = (char*) malloc(sizeof(char) * 30);
    while (fgets(command, 31, fp) != NULL) {
      h = push_history(h, command);
      printf("%s",command);
      interpret_command(command, NULL, c);
    }
    fclose(fp);
    return h;
  }
  return 0;
}

// Interpret and execute a command
//   return value:
//     0, normal commands such as "line"
//     1, unknown or special commands (not recorded in history[])
//     2, quit
int interpret_command(const char *command, History *h, Canvas *c) {
  char buf[bufsize];
  strcpy(buf, command);
  buf[strlen(buf) - 1] = 0; // remove the newline character at the end

  const char *s = strtok(buf, " ");

  // The first token corresponds to command
  if (strcmp(s, "line") == 0) {
    int x0, y0, x1, y1;
    x0 = 0; y0 = 0; x1 = 0; y1 = 0; // initialize
    x0 = atoi(strtok(NULL, " "));
    y0 = atoi(strtok(NULL, " "));
    x1 = atoi(strtok(NULL, " "));
    y1 = atoi(strtok(NULL, " "));
    draw_line(c,x0, y0, x1, y1);
    return 0;
  }

  if (strcmp(s, "rect") == 0) {
    int x0, y0, x1, y1;
    x0 = 0; y0 = 0; x1 = 0; y1 = 0; // initialize
    x0 = atoi(strtok(NULL, " "));
    y0 = atoi(strtok(NULL, " "));
    x1 = atoi(strtok(NULL, " "));
    y1 = atoi(strtok(NULL, " "));
    draw_rectangle(c,x0,y0,x1,y1);
    return 0;
  }

  if (strcmp(s, "circle") == 0) {
    int x0, y0, r;
    x0 = 0; y0 = 0; r = 0;
    x0 = atoi(strtok(NULL, " "));
    y0 = atoi(strtok(NULL, " "));
    r = atoi(strtok(NULL, " "));
    draw_circle(c,x0,y0,r);
    return 0;
  }
  
  if (strcmp(s, "save") == 0) {
    s = strtok(NULL, " ");
    save_history(s, h);
    return 1;
  }

  if (strcmp(s, "undo") == 0) {
    reset_canvas(c);
    // do implementation of iterative interpret_command
    //
    h = undo(h,c);
    if (h == NULL){
      return 3;
    }
    return 1;
  }

  if (strcmp(s, "load") == 0) {
    return 4;
  }

  if (strcmp(s, "history") == 0) {
    printf("Showing History\n");
    show_history(h);
    return 1;
  }

  if (strcmp(s, "pen") == 0) {
    char *pen_type = strtok(NULL, " ");
    if (pen_type == NULL) {
      c->pen = '#';
    } else {
      c->pen = pen_type[strlen(pen_type)-1];
    }
    // char *color = strtok(NULL, " ");
    return 0;
  }

  if (strcmp(s, "quit") == 0) {
    return 2;
  }

  printf("error: unknown command.\n");

  return 1;
}

int main() {
  const int width = 150;
  const int height = 40;
  char pen = '*';
  
  FILE *fp;
  char buf[bufsize];
  fp = stdout;
  Canvas *c = init_canvas(width,height, pen);
  
  int hsize = 0;  // history size
  int i = 0;
  History *h;
  h = NULL;

  while (1) {
    hsize = history_size(h);

    printf("%d:%d > ", hsize, i);
    fgets(buf, bufsize, stdin);
    if(strlen(buf) == 1) continue;

    const int r = interpret_command(buf,h,c);
    if (r == 2) break;
    if (r == 0) {
      h = push_history(h, buf);
    }
    if (r == 1) {
    }
    if (r == 4) {
      h = load(buf,h,c);
    }

    print_canvas(fp, c);
    // rewind_screen(fp, height+3); // rewind the screen to command input
    // clear_command(fp); // clear the previous command
    i++;
  }

  clear_screen(fp);
  free_canvas(c);
  fclose(fp);

  return 0;
}
